src/diffusion_for_multi_scale_molecular_dynamics/models/loss.py

```diff
@@ -149,6 +159,197 @@ def calculate_unreduced_loss(
+        # q(a_t | a_0) = a_0 \bar{Q}_t a_t^T
+        q_at_bar_a0 = compute_q_xt_bar_xo(one_hot_real_atom_types, q_bar_matrices)
+        q_at_bar_a0 = einops.einsum(
+            q_at_bar_a0, one_hot_noisy_atom_types.float(), "... i , ... i -> ..."
+        )
+
+        # dimension of q_at_bar_a0: batch_size, number_of_atoms
+        posterior_q = (
+            q_at_bar_atm1 * q_atm1_bar_a0 / q_at_bar_a0.unsqueeze(-1).clip(min=self.eps)
+        )  # clip at eps
+        # the unsqueeze in the denominator is to allow a broadcasting
+        # posterior q has dimension: batch_size, number_of_atoms, num_type_atoms
+
+        # we now need to compute p_\theta(a_{t-1} | a_t) using
+        # p_\theta(a_{t-1} | a_t) \propto \sum_{\tilde{a}_0} q(a_{t-1}, a_t | \tilde{a}_0)p_\theta(\tilde{a}_0, a_t)
+        # \propto \sum_{\tilde{a}_0} a_t Q_t^T \circ \tilde{a}_0 \bar{Q}_{t-1} \circ p_\theta(\tilde{a}_0 | a_t)
+        # this is equivalent to doing a_t Q_t^T \circ \bar{Q}_{t-1} p_\theta(a_t)
+        # with a matrix multiplication in the last step
+        # we add a softmax to convert the predictions to normalized probabilities
+        p_atm1_at = self.get_p_atm1_at(
+            predicted_unnormalized_probabilities, q_at_bar_atm1, q_bar_tm1_matrices
+        )
+
+        # get the KL divergence between posterior and predicted prob
+        # do not reduce (average) yet as we will replace the samples with t=1 with a NLL loss
+        # input of kl_div should be log-probabilities - we add eps to avoid log(0)
+        kl_loss = torch.nn.functional.kl_div(
+            torch.log(p_atm1_at + self.eps), posterior_q, reduction="none"
+        )
+        return kl_loss
+
+    @staticmethod
+    def get_p_atm1_at(
+        predicted_unnormalized_probabilities: torch.Tensor,
+        q_at_bar_atm1: torch.Tensor,
+        q_bar_tm1_matrices: torch.Tensor,
+    ) -> torch.Tensor:
+        r"""Compute p(a_{t-1} | a_t).
+
+        .. math::
+            p_\theta(a_{t-1} | a_t) \propto \sum_{\tilde{a}_0} q(a_{t-1}, a_t | \tilde{a}_0)p_\theta(\tilde{a}_0, a_t)
+
+        Args:
+            predicted_unnormalized_probabilities: output of the score network estimating an unnormalized
+                :math:`p(a_0 | a_t)` of dimension [batch_size, number_of_atoms, num_type_atoms] where num_type_atoms
+                includes the MASK token
+            q_at_bar_atm1: conditional posterior :math: `q(a_t | a_{t-1}, a0)` as a tensor with dimension
+                [batch_size, number_of_atoms, num_type_atoms]
+            q_bar_tm1_matrices: one-shot transition matrices at previous step :math:`\bar{Q}_{t-1}` of dimension
+                [batch_size, number_of_atoms, num_type_atoms, num_type_atoms]. An identity matrix is used for t=0.
+
+        Returns:
+            one-step transition normalized probabilities of dimension [batch_size, number_of_atoms, num_type_atoms]
+        """
+        p_atm1_at = q_at_bar_atm1 * einops.einsum(
```

Comment:
I don't think this is quite right. There's a normalization constant missing I think. We should discuss to align on exactly what needs to be implemented.